File: web/vuejs/src/components/shared/transformation.ts
/**
 * Copyright (c) 2025 worldiety GmbH
 *
 * This file is part of the NAGO Low-Code Platform.
 * Licensed under the terms specified in the LICENSE file.
 *
 * SPDX-License-Identifier: Custom-License
 */
import { cssLengthValue } from '@/components/shared/length';
import { Transformation } from '@/shared/proto/nprotoc_gen';

export function transformationCSS(transformation?: Transformation): string[] {
	if (!transformation) {
		return [];
	}

	const transformations: string[] = [];

	if (transformation.rotateZ) {
		transformations.push('rotateZ(' + transformation.rotateZ + 'deg)');
	}

	if (transformation.scaleX) {
		transformations.push('scaleX(' + transformation.scaleX + ')');
	}

	if (transformation.scaleY) {
		transformations.push('scaleY(' + transformation.scaleY + ')');
	}

	if (transformation.scaleZ) {
		transformations.push('scaleZ(' + transformation.scaleZ + ')');
	}

	if (transformation.translateX) {
		transformations.push('translateX(' + cssLengthValue(transformation.translateX) + ')');
	}

	if (transformation.translateY) {
		transformations.push('translateY(' + cssLengthValue(transformation.translateY) + ')');
	}

	if (transformation.translateZ) {
		transformations.push('translateZ(' + cssLengthValue(transformation.translateZ) + ')');
	}

	return [`transform: ${transformations.join(' ')}`];
}

File: presentation/ui/transformation.go
// Copyright (c) 2025 worldiety GmbH
//
// This file is part of the NAGO Low-Code Platform.
// Licensed under the terms specified in the LICENSE file.
//
// SPDX-License-Identifier: Custom-License

package ui

import "go.wdy.de/nago/presentation/proto"

type Transformation struct {
	TranslateX Length
	TranslateY Length
	TranslateZ Length
	ScaleX     float64
	ScaleY     float64
	ScaleZ     float64
	// RotateZ defines rotation in degree
	RotateZ float64
}

func (t Transformation) ora() proto.Transformation {
	return proto.Transformation{
		TranslateX: proto.Length(t.TranslateX),
		TranslateY: proto.Length(t.TranslateY),
		TranslateZ: proto.Length(t.TranslateZ),
		ScaleX:     proto.Float(t.ScaleX),
		ScaleY:     proto.Float(t.ScaleY),
		ScaleZ:     proto.Float(t.ScaleZ),
		RotateZ:    proto.Float(t.RotateZ),
	}
}
